### ISSUE TITLE
Some thoughts/propositions on Subscription/All Access in docs and code

### ISSUE THREAD

**Issue by thebigmunch** [simon-weber/gmusicapi#430]
At first, I started with the idea of updating the docs to the more generic Subscription moniker which Google seems to have taken to using in the Google Music UI and support articles. Along with this, I had implemented a property for Mobileclient to return subscription status for an authenticated Mobileclient instance. The intent was to also use this to possibly add short circuit decorators to methods that required a subscription to provide a more informative message to the user.

**However**, not being a subscriber myself beyond when I intend to test something subscription-related for gmusicapi, I was surprised to find that many of the All Access methods in gmusicapi no longer require a subscription to use. The main (only?) things that remain subscription-only are streaming non-uploaded, non-purchased songs and adding them to your library. Also, searching can be used by free accounts, but subscription content is not returned in the result.

So, the intent of the docs changes I planned remains the same, its scope expanded a bit. I propose to:

* Move all methods that no longer require a subscription to appropriate headers
* remove any warnings about CallFailure when not subscribed from methods that no longer require a subscription
* change all references to All Access to the more generic subscribed/subscription terminology Google now uses

As far as the code goes, here are some possibilities I was thinking about/stumbled upon:

* Add an ``is_subscribed`` @property to the Mobileclient class using a call to the ``config`` endpoint that is used to populate the Settings window in the mobile app to get the ``isNautilusUser`` setting used to track subscription status
* Add a ``require_subscription`` decorator to methods that require subscription. This seems less useful now that most things don't actually require a subscription, but it may be nice for future-proofing. Also, some methods don't require a subscription unless someone is doing something that requires one (e.g. ``get_stream_url``).
* Add a ``NoSubscription`` or ``NotSubscribed`` exception to raise in cases where it might be possible to give a more specific, useful error message to users.
* The ``search_all_access`` method is now a misnomer. Not only does Google no longer call the service All Access (in visible UI anyway), but it does not actually require a subscription to use the ``query`` endpoint. The query endpoint seems to return only non-subscription stuff like stations and activities if used by a free account indicated by the ``tier`` URL param and, of course, auth. Perhaps this could be renamed to ``search_google_music`` and the param set based on subscription status or even by an argument. Or perhaps use multiple methods, though this doesn't seem preferable.

Let me know what you think.

**Comment by simon-weber** [comment]
:+1: to docs changes.

> Add an is_subscribed @property

:+1: 

> Add a require_subscription decorator

:+1:, though we may want to cache subscription status to avoid a latency hit. I'm not sure how invalidation should work if we do cache it.

> Add a NoSubscription or NotSubscribed exception

:+1: 

> The search_all_access method is now a misnomer

Agree. I'm thinking rename to `search` and populate the tier automatically via is_subscribed.

Overall, I'm a big fan of these!

**Comment by thebigmunch** [comment]
Caching the property is easy enough. I found [this lib](https://github.com/pydanny/cached-property) which provides handy decorators for the job including one with TTL functionality.

Right now, the implementation sets the attribute ``self._is_subscribed = False`` in __init__, initializes the property ``self.is_subscribed`` (which sets/returns ``self._is_subscribed``) in login, and sets ``self._is_subscribed = False`` and deletes ``is_subscribed`` from the instance __dict__ (the method given for invalidating the cache for the decorator) in logout. ``is_subscribed`` is a ``cached_property_ttl``.

Or in plain English, subscription status is defaulted to False on instantiation, set/initialized during login, reset during logout, and only refreshes if checked *insert TTL here* or longer after the last time.

The questions are:

* Does 10 minutes sound like a good TTL?
* Should we provide a way for users to force a refresh besides re-logging?
  * Should it be a method?
  * Should we make is_subscribed a regular method that takes an argument? (Not a fan. I already think is_authenticated() should be a property of the client instances as it is for the underlying Session : P)
  * Should we just document the method of invalidating the cache used in logout?
  * None of the above?

Honestly, I'm probably being generous when saying .1% of use may require an update in subscription status during the instance's life or the TTL we set. So I think documenting a way to reset the subscription status if necessary would suffice.

**Comment by simon-weber** [comment]
10 minute TTL sounds good :+1: 

> So I think documenting a way to reset the subscription status if necessary would suffice.

Yeah, that sounds reasonable to me. It's definitely an edge case.

**Comment by Tadly** [comment]
IMHO a `validate_subscription_status()` wouldn't hurt.
From a dev perspective, doing a `gmusic.logout(); login.login()` isn't that big of a deal but `validate_subscription_status()` would make *more sense* when reading source.

Though, when there's a `validate_subscription_status()`, suddenly a `is_subscribed()` would make sense which goes against the pure decorator approach...

Personally I think having those to methods would be nice.
Just my two cents though :)

**Comment by thebigmunch** [comment]
I'm not quite sure what use you had in mind for ``validate_subscription_status``, but it has the same meaning as ``is_subscribed`` just longer and uglier. ``is_subscribed`` (not ``is_subscribed()``, btw) will be the property which checks/returns the cached subscription status with a TTL. The question is whether to add a method to force a refresh of the subscription status during that TTL probably with the name ``refresh_subscription`` or similar. But instead, documenting what would be in that method for users to use if necessary is better, since it's unlikely to actually be required and adding a too many methods just makes things worse. Forcing a refresh looks like this:

```
# Invalidate cache and reset subscription status.
del self.__dict__['is_subscribed']
self._is_subscribed = False

# Refresh subscription status.
self.is_subscribed
```

which can easily be turned into a function in user code. That's likely something I'd add to my wrapper, for example. You can even have that function ``return self.is_subscribed`` to return the new subscription status.

**Comment by Tadly** [comment]
Well. Following is messing with the internals of a module/lib (which you, as module user, shouldn't be doing)
```
# Invalidate cache and reset subscription status.
del self.__dict__['is_subscribed']  # not sure what to think about that one. Is this really something you want a library-user to do?
self._is_subscribed = False  # private, shouldn't be touched from outside the lib?

# Refresh subscription status.
self.is_subscribed  # <- refreshing/re-validating the subscription status should be a action (method so to speak) rather than "accessing" something.
```

The "use in mind" was my kodi add-on where I could give the user the ability to re-check there subscription status (if they'r logged in already and don't want to wait for the 10 min. TTL) as I don't use `logout` at all (and login only to make sure the session is still valid).

**BUT** I'm nitpicking here and as long as it's documented on how to re-validate the subscription status, it's fine by me :)

**Comment by thebigmunch** [comment]
Actually, you're changing the internals of a class instance, not a module or lib or even a class, which, by the way, you're already doing for Mobileclient auth token login. And the plan is to document that workaround as well, so this would be consistent.

You'd be doing the same if gmusicapi just hid it behind a method anyway, the user just wouldn't be aware of the internals. This kinda just happens when you cache something and then try to provide a way to break the cache. You're going to have to mess with the internals of something. If you're worried about the UX of refreshing a token, then the better argument might be, "Do we really need to cache it? Just make it give a live result every time", which is how I originally wrote it.

What are the odds that someone will need to refresh it within 10 minutes (or at all)? If they did, what are the odds that they don't just logout/login or start a new instance? Is it worth creating a method for such a rare case? Maybe, but you're still messing with internals even if indirectly : )

"Refresh subscription status" comment could really be "Initialize subscription status". I used refresh to indicate that the instance at some point had a subscription status, but initialize is more correct, since the status is back to default. As for whether private things shouldn't be touched, the axiom is "We're all adults here.". Private attributes/methods are marked to show that they aren't expected to be used outside the class, not that they can't be. They are an implementation detail that users can choose to accept the risk of using despite the fact that it could change without notice as it's not part of the public API (unlikely to happen here since it'd be documented).

**Comment by thebigmunch** [comment]
Thinking about it, I could probably make the UX cleaner by adding a deleter for the ``is_subscribed`` property The user would then just need to do ``del self.is_subscribed`` to handle invalidating the cache and then check self.is_subscribed again. And this is a proper usage pattern for a class property. This is what happens when you don't take a step back from coding to think for awhile; you miss the easy stuff ^_^

**Comment by Tadly** [comment]
Sounds good to me. Way nicer than the initial proposal :)
Good thinking :+1:

**Comment by thebigmunch** [comment]
Side note: There's actually discussion about adding a @cached_property and possibly a TTL version to the standard lib in 3.6, though the current thought is that TTL shouldn't be in the standard lib. Discussion is [here](https://bugs.python.org/issue21145).

**Comment by thebigmunch** [comment]
So, didn't wait until after work : )

Added a different implementation of @cached_property to gmusicapi.utils.utils that, with a small addition, better suits our needs. It opens the door for future cached properties, and could be modified if the need for explicit setters/deleters arises.
